### ISSUE TITLE
[Rejected] SDL 0271 - Start APT of SDL App by pressing PTT button

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#893]
Hello SDL community,

The review of "SDL 0271 - Start APT of SDL App by pressing PTT button"  begins now and runs through December 17, 2019.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0271-Start-APT-of-SDL-App-by-pressing-PTT-Button.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/893

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by atiwari9** [comment]
Overall a good concept, but i have some concerns for the implementation:

1. I am not in favor of defining HMI actions for a Long/Short Press or when HMI should send `PUSH_TO_TALK` button press notification to SDL Core/app. Some OEMs have PTT mapped to phone's native voice assistants like SIRI/Google Assistant etc. HMI should be able to decide the arbitration between Native VR/Phone voice assistant/SDL PTT.

2. I do not think it'd be good user experience if PTT app keeps changing for the user based on HMI Level changes. There has to be clear selection for an app which could provide PTT/Voice Assistance service. We need to think on this point a bit more. I am considering putting a limitation in terms of say apps with HMI type `NAVIGATION/PROJECTION` can't request subscription to `PUSH_TO_TALK` so that we do not have to decide between two apps in HMI Limited, Additionally, as a best practice HMI should notify the user that _xyz app is providing voice assistance now_ , when such app registers and is available to provide this service.

3. As an extension to this proposal, we should also consider registering some **Wake Words** with HMI. We can control wake words allowed for an app to be used via policies as well to ensure that there are no conflicts between apps. When such word is spoken by user, HMI can send `PUSH_TO_TALK` notification to correct app. This approach would solve:
    * Conflict between Hard Key PTT for Native VR vs SDL PTT as unique wake words could be registered.
    * Which app should get `PUSH_TO_TALK` notification,

**Comment by Shohei-Kawano** [comment]
@atiwari9 -san
Thank you for your comment.

1. Do you know if some OEMs maps phone's native voice assistant to short press or long press?
If both long and short presses are mapped, OEMs should choose not to support PTT for SDL or to exclude it from phone's native voice assistant.
The proposal I showed is just a recommendation.
OEM can decide how to mediate VR Service with HMI.
 
2. As you said, it is not good that the operation changes depending on the HMI Level. 
I think the operation should be as constant as possible.

Why does the HMI type NAVIGATION / PROJECTION app require a restriction that it cannot request a subscription to PUSH_TO_TALK?

3. I think this is another way to use Wake up word.
However, I think that it will be another proposal because it needs a mechanism to recognize Wake up word with IVI Native.
I think that it is a good idea for further improvement of UX, but if you include it in this proposal, I think that the scale will become very large. Therefore, I think that it is better to remove this PTT button addition proposal. How is it?

Thanks.

**Comment by joeljfischer** [comment]
While I appreciate the desire for this feature, I don't believe that the current proposal or approach is adequate to support this feature and a completely different approach should be attempted using app services.

I don't believe that the updated proposal solves the major problems of the original.

1. Dealing with the OEM's voice assistant. The proposal states that this will override the the OEM's voice assistant if an app that subscribes to the PTT button is in HMI_FULL and sometimes in HMI_LIMITED. This brings up a bunch of complexity for the user. How does the user know which voice assistant will be activated? What if they always want the OEM voice assistant to be used? What if they want a different voice assistant to be used even if it's in HMI_BACKGROUND? This is all complexity that would have to be handled by the HMI, but has already been solved in the app services system. I don't believe that your chart being merely recommended is enough. These problems need to actually be solved and not left up to each OEM to solve themselves.

2. I'm also concerned by the different behaviors based on HMI level. This seems like it will be unintuitive to the user. I also don't like that HMI_BACKGROUND apps will not have the ability to control the button. If a voice assistant app is in the background, they should be able to control the button if desired.

3. Wake up word support should support multiple wake up words per app and various languages and accents.

4. This doesn't support "hold to talk." This might be fine, but it is a deficiency mentioned by Amazon in the previous proposal's comments.

5. This doesn't support multi-talk interactions, e.g. user: "play some music", app: "what music would you like to play, classical or rock?", user: "classical", app: "playing classical music"

Overall, I don't believe that this proposal adequately addresses the issues with the previous version of the proposal and doesn't support necessary features for voice assistants. I think that we should move away from the solution of using a button event and reject this proposal.

We had previous discussions on voice assistant support via app services. I think that is a much better potential solution because it would theoretically solve (1) and (2) via the existing app services system, and it should easily solve (3) if it is properly done. (4) and (5) should be solvable using this scheme as well with additional RPCs.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
Thank you for your comment.

I think a good UX will change depending on the OEM concept.
I believe that SDL Core needs to provide a mechanism that can realize UX that OEMs think is good.
I think my proposal is one of them.
Of course, Wakeup Word that you propose is one of them.

Presenting some recommended UXes is very useful for IVI development, so it's a good idea to present them as examples.

In other words, I think it's better to discuss the method, not the good UX.
How is it?

Thanks.

**Comment by joeygrover** [comment]
>In other words, I think it's better to discuss the method, not the good UX.

I appreciate you want to create a method that works for your integration, however I disagree with this statement. SDL is a technology that is being used as a brand. If we don't define a good, stable UX as a consortium that brand suffers. We are also not talking about just end-user experience, but Core integrator and app developer experiences as well. The proposal as it currently stands does not work well for any of those parties. It contains many shortcomings that @joeljfischer has provided in his comment. As was with the original proposal that this one is based off from, this feature does not work well with the method provided. It creates many different scenarios that must be covered by the SDL projects, app developers, and core integrators. With so many possibilities it creates a fragile feature that is impossible to properly test and make robust. 

The path forward here is to use the app services feature. We were expecting @AndrewRMitchell to provide a more feature rich version of the proposed solution initially, however I think we can create a more basic version first to get integrated and satisfy the needs that this proposal is attempting to solve. You can see the first draft [here](https://github.com/smartdevicelink/sdl_evolution/issues/599). 

My recommendation is to take the very basic pieces of #599 and extend it slightly. An idea would be to create a new capability param with a new `VoiceAssistantOperatingMode` enum with values of `AUDIO_PASSTHRU`and `MICROPHONE_DIRECT` to start. The param would be an array of modes supported. When a voice session is activated (via wake word, PTT button, etc), the active voice assistant(VA) service would then be selected to receive the incoming data based on their supported operating modes. For example,  if the VA had only `AUDIO_PASSTHRU` contained in the operating mode the raw voice data would be passed to the VA  via an audio pass thru session and they would then perform actions on it. The `MICROPHONE_DIRECT` capability would be if the app can obtain data directly from the vehicles microphone with something like bluetooth hands free profile, reducing latency of processing data into and out of SDL packets. I think other modes could be added later, but these two would support a basic VA activation and processing. With the management of active services already defined by the app services feature, the UX and methods would be much improved over the solution provided by this proposal.

**Comment by Shohei-Kawano** [comment]
I wrote a misleading sentence. I'm sorry...

My opinion was not to discuss which is the best UX, but to discuss what is the way to achieve a good UX.
I think it is necessary to discuss various UX and decide which UX you want to achieve.
However, I think it is very difficult to implement functions that satisfy all UXes at once.
Therefore, I think it is necessary to decide the priority order and draw the road map.

Maybe my proposal may be the first step. maybe.
It may be diffrent...

I did not deep-read the proposal of [#599](https://github.com/smartdevicelink/sdl_evolution/issues/599) yet.
First of all, I need to deep-read it, so wait a moment.

**Comment by joeygrover** [comment]
Understood, and  I agree it is hard to create a _perfect_ solution from the start. I think starting with app services gives us the best chance to create a basic feature for this to be added soon that can then be easily improved in the future updates. To fully understand #599, you must also read #485 for the full app service framework. Voice Assistance is just a potential service type that uses the feature of app services. It is a long proposal and I can happily answer any questions you have about it or how this feature could potentially work with it.

**Comment by theresalech** [comment]
The Steering Committee voted to reject this proposal.  It is rejected in favor of a new proposal to extend the App Services feature to include voice assistant functionality.  The author of this proposal will review the App Services feature in more detail, including #599 and #485 and submit a new proposal to utilize that solution when ready.
